### PR TITLE
xtensa/esp32: Add SPI Flash device driver

### PR DESCRIPTION
**Pull request by acassis**
ESP32 runs code in a SPI Flash, so users can also use it to store
data directly or mount some parts into a filesystem.

The SPI Flash usually use SPI0.

This driver was implemented by Dong Heng <dongheng@espressif.com>
and modified to fix coding style by Alan Carvalho de Assis.

### CODE REVIEW

File: arch/xtensa/src/esp32/rom/esp32_spiflash.h [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+{
+#endif
+
+/** \defgroup spi_flash_apis, spi flash operation related apis
+  * @brief spi_flash apis
+  */
+
+/** @addtogroup spi_flash_apis
+  * @{
+  */
```
Comment (patacongo) [inline arch/xtensa/src/esp32/rom/esp32_spiflash.h:42]: Use of Doxygen tags is specifically forbidden by the NuttX coding standard.  Please remove lines 36-42 and 815-817.

File: arch/xtensa/src/esp32/rom/esp32_spiflash.h [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+
+typedef struct
+{
+    uint8_t  data_length;
+    uint8_t  read_cmd0;
+    uint8_t  read_cmd1;
+    uint8_t  write_cmd;
+    uint16_t data_mask;
+    uint16_t data;
+} esp_rom_spiflash_common_cmd_t;
```
Comment (patacongo) [inline arch/xtensa/src/esp32/rom/esp32_spiflash.h:174]: All of the about typedefs belong in a "Public Types", not under "Pre-processor Definitions"

File: arch/xtensa/src/esp32/rom/esp32_spiflash.h (hunk 1 above)
Comment (acassis) [reply arch/xtensa/src/esp32/rom/esp32_spiflash.h:42]: Ok, I changed almost all Doxygen to NuttX standard, but these two were missed. Thank you Greg!

File: arch/xtensa/src/esp32/rom/esp32_spiflash.h (hunk 2 above)
Comment (acassis) [reply arch/xtensa/src/esp32/rom/esp32_spiflash.h:172]: Thank you Greg!

**Pull request merged**